File: gunhoflash/spiral-matrix.py
from typing import List

class Solution:
  def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
    startRow = 0
    endRow = len(matrix) - 1
    startCol = 0
    endCol = len(matrix[0]) - 1

    answer = []
    while startRow <= endRow and startCol <= endCol:
      if startRow == endRow:
        for col in range(startCol, endCol + 1):
          answer.append(matrix[startRow][col])
        break
      if startCol == endCol:
        for row in range(startRow, endRow + 1):
          answer.append(matrix[row][endCol])
        break
      for col in range(startCol, endCol):
        answer.append(matrix[startRow][col])
      for row in range(startRow, endRow):
        answer.append(matrix[row][endCol])
      for col in range(endCol, startCol, -1):
        answer.append(matrix[endRow][col])
      for row in range(endRow, startRow, -1):
        answer.append(matrix[row][startCol])
      startRow += 1
      endRow -= 1
      startCol += 1
      endCol -= 1

    return answer

File: gunhoflash/spiral-matrix.ts
function spiralOrder(matrix: number[][]): number[] {
  let startRow = 0;
  let endRow = matrix.length - 1;
  let startCol = 0;
  let endCol = matrix[0].length - 1;

  const answer = Array((endRow - startRow + 1) * (endCol - startCol + 1)).fill(0);
  let i = 0;
  while (startRow <= endRow && startCol <= endCol) {
    if (startRow === endRow) {
      for (let col = startCol; col <= endCol; col++) {
        answer[i++] = matrix[startRow][col];
      }
      break;
    }
    if (startCol === endCol) {
      for (let row = startRow; row <= endRow; row++) {
        answer[i++] = matrix[row][endCol];
      }
      break;
    }
    for (let col = startCol; col < endCol; col++) {
      answer[i++] = matrix[startRow][col];
    }
    for (let row = startRow; row < endRow; row++) {
      answer[i++] = matrix[row][endCol];
    }
    for (let col = endCol; col > startCol; col--) {
      answer[i++] = matrix[endRow][col];
    }
    for (let row = endRow; row > startRow; row--) {
      answer[i++] = matrix[row][startCol];
    }
    startRow++;
    endRow--;
    startCol++;
    endCol--;
  }

  return answer;
}
